### PR TITLE
S3 buckets

### PR DESCRIPTION
**Pull request by benvand**
Terraform for preview staging and production s3 buckets.

See [README.md](https://github.com/alphagov/digitalmarketplace-aws/pull/408/files#diff-e3ac418a4fe3b69c3e62825d6d1001d1) for more info

### CODE REVIEW

**Comment by katstevens** [comment]
Can we put a line in the S3 section of the manual (https://alphagov.github.io/digitalmarketplace-manual/aws-resources.html#s3-buckets) to say which role you need to switch to for each environment, in order to view the buckets defined here? I was double checking what we had, and before I'd switched role I could view a whole bunch of old buckets (and the public bucket, which we definitely need for at least one Jenkins script) that I don't think are defined anywhere in this repo.

**Comment by benvand** [comment]
@katstevens 
> Do we need to explicitly configure ACL as part of the policy, or should we update the Cloudtrail stuff to match this?

Yeah we should update the old style definition to use computed IAM policy document data sources. It doesn't actually make any difference to the end product but the policy document style is way more readable.

> Can we put a line in the S3 section of the manual (https://alphagov.github.io/digitalmarketplace-manual/aws-resources.html#s3-buckets) to say which role you need to switch to for each environment, in order to view the buckets defined here?

Sure, we'll need to keep this up to date when we change them though. I'll see if I can link to the relevant groups in creds 🤷‍♀️

**Comment by benvand** [comment]
Also re the old buckets, yeah, they're in the main account, and probably shouldn't be, but they're out of scope of this PR (But we should do something about them soon.)

File: terraform/environments/preview/s3_buckets.tf [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    target_bucket = "${aws_s3_bucket.server_access_logs_bucket.id}"
+    target_prefix = "digitalmarketplace-documents-preview-preview/"
+  }
+
+  policy = "${data.aws_iam_policy_document.documents_bucket_policy_document.json}"
+}
+
+# G7-draft-documents
+
+resource "aws_s3_bucket" "g7-draft-documents_bucket" {
```
Comment (Wynndow) [inline terraform/environments/preview/s3_buckets.tf]: No policy for this bucket?

File: terraform/environments/preview/s3_buckets.tf [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+  logging {
+    target_bucket = "${aws_s3_bucket.server_access_logs_bucket.id}"
+    target_prefix = "digitalmarketplace-g7-draft-documents-preview-preview/"
+  }
+}
+
+# Submissions
+
+resource "aws_s3_bucket" "submissions_bucket" {
```
Comment (Wynndow) [inline terraform/environments/preview/s3_buckets.tf]: No policy on this bucket either?

File: terraform/environments/preview/s3_buckets.tf [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  versioning {
+    enabled = true
+  }
+
+  logging {
+    target_bucket = "${aws_s3_bucket.server_access_logs_bucket.id}"
+    target_prefix = "digitalmarketplace-agreements-preview-preview/"
+  }
+
+  policy = "${data.aws_iam_policy_document.agreements_bucket_policy_document.json}"
```
[the remark this answers is not in the document]
Comment (benvand) [reply terraform/environments/preview/s3_buckets.tf]: Yeah, pretty nifty. Better than writing plain json

File: terraform/environments/preview/s3_buckets.tf (hunk 1 above)
Comment (benvand) [reply terraform/environments/preview/s3_buckets.tf]: screen shot 2018-07-26 at 15 36 41

Comment (benvand) [reply terraform/environments/preview/s3_buckets.tf]: screen shot 2018-07-26 at 15 36 58

Comment (benvand) [reply terraform/environments/preview/s3_buckets.tf]: Nada 🤔  Bit weird

File: terraform/environments/preview/s3_buckets.tf (hunk 2 above)
Comment (benvand) [reply terraform/environments/preview/s3_buckets.tf]: screen shot 2018-07-26 at 15 36 41

Same

**Pull request merged**